File: app/src/main/java/com/example/vladimir/easyenglishlearn/exercises/WordQuizFragment.java
package com.example.vladimir.easyenglishlearn.exercises;


import androidx.lifecycle.ViewModelProvider;
import androidx.databinding.DataBindingUtil;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.fragment.app.Fragment;

import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.example.vladimir.easyenglishlearn.R;
import com.example.vladimir.easyenglishlearn.databinding.FragmentWordQuizBinding;
import com.example.vladimir.easyenglishlearn.model.Word;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

import static com.example.vladimir.easyenglishlearn.Constants.SELECTED_WORDS;
import static com.example.vladimir.easyenglishlearn.Constants.TRANSLATION_DIRECTION;

public class WordQuizFragment extends Fragment {

    private FragmentWordQuizBinding mBinding;


    @NonNull
    public static Fragment newInstance(ArrayList<Word> selectedWordList,
                                       boolean translationDirection) {
        Bundle args = new Bundle();
        args.putParcelableArrayList(SELECTED_WORDS, selectedWordList);
        args.putBoolean(TRANSLATION_DIRECTION, translationDirection);
        Fragment fragment = new WordQuizFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    @SuppressWarnings("ConstantConditions")
    public View onCreateView(@NonNull LayoutInflater inflater,
                             ViewGroup container,
                             Bundle savedInstanceState) {
        mBinding = DataBindingUtil.inflate(inflater,
                R.layout.fragment_word_quiz,
                container,
                false);
        WordQuizViewModel viewModel = new ViewModelProvider(this).get(WordQuizViewModel.class);
        mBinding.setViewModel(viewModel);
        subscribeToLiveData(viewModel);

        if (savedInstanceState == null) {
            boolean translationDirection = getArguments().getBoolean(TRANSLATION_DIRECTION);
            List<Word> wordList = getArguments().getParcelableArrayList(SELECTED_WORDS);
            viewModel.startExercise(wordList, translationDirection);
        }
        return mBinding.getRoot();
    }

    private void subscribeToLiveData(@NonNull WordQuizViewModel viewModel) {
        viewModel.getExerciseCloseLiveData().observe(getViewLifecycleOwner(), aVoid -> closeFragment());
        viewModel.getMessageLiveData().observe(getViewLifecycleOwner(), this::showMessage);
        viewModel.getClearRadioGroupLiveData().observe(getViewLifecycleOwner(), aVoid -> clearRadioGroup());
    }

    private void clearRadioGroup() {
        mBinding.wqfRgAnswers.clearCheck();
    }

    private void closeFragment() {
        Objects.requireNonNull(getActivity()).onBackPressed();
    }

    private void showMessage(int errorsCount) {
        String message;
        if (errorsCount < 0) {
            message = getString(R.string.wrong_answer);
        } else {
            message = getString(R.string.errors_count, errorsCount);
        }
        Toast.makeText(getActivity(), message, Toast.LENGTH_SHORT).show();
    }
}

File: app/src/main/java/com/vladimir_x/easyenglishlearn/exercises/WordQuizFragment.kt
package com.vladimir_x.easyenglishlearn.exercises

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import com.vladimir_x.easyenglishlearn.Constants
import com.vladimir_x.easyenglishlearn.R
import com.vladimir_x.easyenglishlearn.databinding.FragmentWordQuizBinding
import com.vladimir_x.easyenglishlearn.model.Word
import java.util.ArrayList

class WordQuizFragment : Fragment() {
    private var _binding: FragmentWordQuizBinding? = null
    private val binding get() = _binding!!

    lateinit var viewModel: WordQuizViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = FragmentWordQuizBinding.inflate(
            inflater,
            container,
            false
        )
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        viewModel = ViewModelProvider(this)[WordQuizViewModel::class.java]
        subscribeToLiveData(viewModel)
        if (savedInstanceState == null) {
            val translationDirection =
                arguments?.getBoolean(Constants.TRANSLATION_DIRECTION) ?: true
            val wordList: List<Word> =
                arguments?.getParcelableArrayList<Word>(Constants.SELECTED_WORDS) as List<Word>
            viewModel.startExercise(wordList, translationDirection)
        }
    }

    private fun subscribeToLiveData(viewModel: WordQuizViewModel) {
        with(viewModel) {
            exerciseCloseLiveData.observe(viewLifecycleOwner) {
                closeFragment()
            }
            messageLiveData.observe(viewLifecycleOwner) {
                it?.let { errorsCount: Int ->
                    showMessage(errorsCount)
                }
            }
            clearRadioGroupLiveData.observe(viewLifecycleOwner) {
                clearRadioGroup()
            }
            questionLiveData.observe(viewLifecycleOwner, ::fillFields)
        }

    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun clearRadioGroup() {
        _binding!!.wqfRgAnswers.clearCheck()
    }

    private fun closeFragment() {
        requireActivity().onBackPressed()
    }

    private fun showMessage(errorsCount: Int) {
        val message: String = if (errorsCount < 0) {
            getString(R.string.wrong_answer)
        } else {
            getString(R.string.errors_count, errorsCount)
        }
        Toast.makeText(activity, message, Toast.LENGTH_SHORT).show()
    }

    private fun fillFields(pair: Pair<String, List<String>>) {
        val answers = pair.second
        with(binding) {
            wqfTvQuestion.text = pair.first
            wqfRbFirst.text = answers[0]
            wqfRbFirst.setOnClickListener { viewModel.onAnswerChecked(0) }
            wqfRbSecond.text = answers[1]
            wqfRbSecond.setOnClickListener { viewModel.onAnswerChecked(1) }
            wqfRbThird.text = answers[2]
            wqfRbThird.setOnClickListener { viewModel.onAnswerChecked(2) }
        }
    }

    companion object {
        fun newInstance(
            selectedWordList: ArrayList<Word>?,
            translationDirection: Boolean
        ) = WordQuizFragment().apply {
            arguments = createBundle(selectedWordList, translationDirection)
        }

        private fun createBundle(
            selectedWordList: ArrayList<Word>?,
            translationDirection: Boolean
        ) = Bundle().apply {
            putParcelableArrayList(Constants.SELECTED_WORDS, selectedWordList)
            putBoolean(Constants.TRANSLATION_DIRECTION, translationDirection)
        }
    }
}